packages/react-dom-bindings/src/client/ReactDOMHostConfig.js

```diff
@@ -1976,3 +1762,1541 @@ export function clearSingleton(instance: Instance): void {
   }
   return;
 }
+
+// -------------------
+//     Resources
+// -------------------
+
+export const supportsResources = true;
+
+// The resource types we support. currently they match the form for the as argument.
+// In the future this may need to change, especially when modules / scripts are supported
+type ResourceType = 'style' | 'font' | 'script';
+
+type HoistableTagType = 'link' | 'meta' | 'title';
+type TResource<
+  T: 'stylesheet' | 'style' | 'script' | 'void',
+  S: null | {...},
+> = {
+  type: T,
+  instance: null | Instance,
+  count: number,
+  state: S,
+};
+type StylesheetResource = TResource<'stylesheet', StylesheetState>;
+type StyleTagResource = TResource<'style', null>;
+type StyleResource = StyleTagResource | StylesheetResource;
+type ScriptResource = TResource<'script', null>;
+type VoidResource = TResource<'void', null>;
+type Resource = StyleResource | ScriptResource | VoidResource;
+
+type LoadingState = number;
+const NotLoaded = /*       */ 0b00;
+const Loaded = /*          */ 0b01;
+const Errored = /*         */ 0b10;
+const Settled = /*         */ 0b11;
```

Comment:
Nit: Seems like this should just be an enum instead.

Reply:
I'm rethinking this part and will take another pass in a follow up PR.

I added another state `Inserted` in a recent refactor so I'll leave as a bitmask for now. I think I want to actually rework the whole thing to use something like a thenable with status embedded. I was thinking for some reason that the load event handlers might run sync sometimes but I'm pretty sure that's just wrong and they'll always just be in a separate task so using a thenable could actually have the commit happen sooner in a microtask